Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +352,141 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
-    mlmg.setVerbose(2);
+    mlmg.setVerbose(0);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    geometry.potential_lo and geometry.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setDirichletBC(amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                      const int idim) const
+{
+    Real potential_lo;
+    Real potential_hi;
+
+    getBoundaryPotentials(idim, potential_lo, potential_hi);
+
+    // Print() << "Potential lo = " << potential_lo << "\n";
+    // Print() << "Potential hi = " << potential_hi << "\n";
+
+    // loop over all multigrid levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+        for (MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+            FArrayBox& fab = (*phi[0])[mfi]; //
+            Array4<Real> const& phi_gh = fab.array(); // phi include ghost cells
+            const auto lo = lbound(phi_gh);
+            const auto hi = ubound(phi_gh);
+
+            //TODO use AMREX_GPU_DEVICE for acceleration
+            // The boundary values are set at points lo.idim + 1 and hi.idim -1
+            // since phi_gh includes ghost cells
+            if (idim == 0){
+                for (int k=lo.z; k<=hi.z; k++) {
```

Comment:
Can you write these loops using the ParallelFor construct? Also, this doesn't take into account the block structure of the data. The domain will be divided up into blocks and only blocks at the edge of the domain would need the potential set on the boundary. Take a look, for example, at how this is done in PR#1767. You should consider rearranging the code so that all of the boundaries are set in a single call rather than calling this routine for each axis.

Reply:
I've incorporated the ParallelFor construct (which is definitely a better approach) and rearranged things so that all boundaries are set in a single call. 
I'm not quite sure if the issue of only looping over blocks on the domain edges have been taken care of. I looked at PR#1767 as you suggested and tried to follow the logic in `PEC::ApplyPECtoEfield`. What I make out from it is that there is an outer loop over the AMR levels, an inner loop over boxes and then the ParallelFor to loop over tiles. Inside the ParallelFor the tiles are checked to see if they hold boundary cells, and if so those are set. Did I miss anything in that which would make the picking out of edge blocks more efficient?